src/Razor/src/Microsoft.NET.Sdk.Razor/DiscoverMvcApplicationParts.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.AspNetCore.Razor.Tasks
+{
+    public class DiscoverMvcApplicationParts : Task
+    {
+        [Required]
+        public ITaskItem[] MvcAssemblyNames { get; set; }
+
+        [Required]
+        public ITaskItem[] ResolvedReferences { get; set; }
+
+        [Output]
+        public string[] ApplicationPartAssemblyNames { get; set; }
+
+        public override bool Execute()
+        {
+            var referenceItems = new List<ResolveReferenceItem>();
+            foreach (var item in ResolvedReferences)
+            {
+                string assemblyName = null;
+                var fusionName = item.GetMetadata("FusionName");
+                if (!string.IsNullOrEmpty(fusionName))
+                {
+                    assemblyName = new AssemblyName(fusionName).Name;
+                }
+                else
+                {
+                    assemblyName = GetAssemblyName(item.ItemSpec);
+                }
+
+                if (assemblyName == null)
+                {
+                    continue;
+                }
+
+                referenceItems.Add(new ResolveReferenceItem
+                {
+                    AssemblyName = assemblyName,
+                    IsSystemReference = item.GetMetadata("IsSystemReference") == "true",
+                    Path = item.ItemSpec,
+                });
+            }
+
+            var mvcAssemblyNames = MvcAssemblyNames.Select(s => s.ItemSpec).ToList();
+
+            var provider = new ApplicationPartsProvider(mvcAssemblyNames, referenceItems);
+            var assemblyNames = provider.ResolveAssemblies();
+
+            ApplicationPartAssemblyNames = assemblyNames.Count > 0 ?
+                assemblyNames.ToArray() :
+                Array.Empty<string>();
+            return true;
```

Comment:
Style nit: 
```suggestion
            return !Log.HasLoggedErrors;
```

This prevents the "I logged an error but returned success, causing the build to succeed but with logged errors" case.